### PR TITLE
[FLINK-20304] Introduce the BroadcastStateTransformation.

### PR DESCRIPTION
**Pull request by kl0u**
## What is the purpose of the change

So far the Broadcast State pattern was a TwoInputTransformation.
This did not allow for special handling when it comes to
translating it for Batch or Streaming execution. This commit
introduces a special transformation just for this.

## Brief change log

We introduce the `BroadcastStateTransformation` and the corresponding `BroadcastStateTransformationTranslator` that throws an `UnsupportedOperationException` when called for execution in `BATCH` mode.

## Verifying this change

The `STREAMING` translation is covered by existing tests while I added the `BroadcastStateITCase.testBroadcastBatchTranslationThrowsException()` for the `BATCH`case.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java [hunk 1]
```diff
@@ -57,8 +58,8 @@
 public class BroadcastConnectedStream<IN1, IN2> {
 
 	private final StreamExecutionEnvironment environment;
-	private final DataStream<IN1> inputStream1;
-	private final BroadcastStream<IN2> inputStream2;
+	private final DataStream<IN1> nonBroadcastStream;
```
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java:61]: I really appreciate the renaming!

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/BroadcastStateTransformation.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+				parallelism);
+	}
+
+	private BroadcastStateTransformation(
+			final String name,
+			final DataStream<IN1> inputStream,
+			final BroadcastStream<IN2> broadcastStream,
+			final StreamOperatorFactory<OUT> operatorFactory,
+			final TypeInformation<?> keyType,
+			final KeySelector<IN1, ?> keySelector,
```
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/BroadcastStateTransformation.java:106]: add `@Nullable` annotations

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/BroadcastStateTransformation.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+				checkNotNull(inputStream),
+				broadcastStream,
+				operatorFactory,
+				null,
+				null,
+				outTypeInfo,
+				parallelism);
+	}
+
+	public BroadcastStateTransformation(
```
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/BroadcastStateTransformation.java:82]: I am not the biggest fan of having multiple constructors. Having a single entry point to a class makes it easier to reason about its fields state. Could we change the ctor to static methods? Or even just expose the actual ctor and move the keyselector extraction to the `KeyedStream`?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/BroadcastStateTransformation.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+ *     Broadcast State Pattern documentation page</a>.
+ *
+ * @param <IN1> The type of the elements in the non-broadcasted input.
+ * @param <IN2> The type of the elements in the broadcasted input.
+ * @param <OUT> The type of the elements that result from this transformation.
+ */
+@Internal
+public class BroadcastStateTransformation<IN1, IN2, OUT> extends PhysicalTransformation<OUT> {
+
+	private final DataStream<IN1> inputStream;
```
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/BroadcastStateTransformation.java:54]: How about keeping `Transformations` here.

The way it is now creates an architectural cyclic dependency: DataStream is an abstraction on top of Transformation, but BroadcastStateTransformation keeps DataStream as a field.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/BroadcastStateTransformationTranslator.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+	@Override
+	protected Collection<Integer> translateForBatchInternal(
+			final BroadcastStateTransformation<IN1, IN2, OUT> transformation,
+			final Context context) {
+		throw new UnsupportedOperationException("The Broadcast State Pattern is not support in BATCH execution mode.");
+	}
+
+	@Override
+	protected Collection<Integer> translateForStreamingInternal(
+			final BroadcastStateTransformation<IN1, IN2, OUT> transformation,
```
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/BroadcastStateTransformationTranslator.java:55]: Could we unify this between `BroadcastStateTransformationTranslator` and `TwoInputTransformationTranslator`? The entire method is identical in the two classes (minus checking for `keySelector` in a single input, but the version from `TwoInputTransformationTranslator` should work as well.)

We did a similar thing with `AbstractOneInputTransformationTranslator`.

File: flink-tests/src/test/java/org/apache/flink/test/streaming/runtime/BroadcastStateITCase.java [hunk 6]
```diff
@@ -149,6 +153,45 @@ public long extractTimestamp(String element, long previousElementTimestamp) {
...
+
+					@Override
+					public void processBroadcastElement(Long value, Context ctx, Collector<Long> out) {
+
+					}
+				});
+
+		try {
+			env.execute();
+			fail("Execution should have failed during job graph creation.");
```
Comment (dawidwys) [inline flink-tests/src/test/java/org/apache/flink/test/streaming/runtime/BroadcastStateITCase.java:189]: How about we use the `ExpectedException` rule instead? I find it more concise. 

```
@Rule
public ExpectedException thrown = ExpectedException.none();

thrown.expect(UnsupportedOperationException.class);
thrown.expectMessage("The Broadcast State Pattern is not support in BATCH execution mode.");
```

**Review by dawidwys — CHANGES_REQUESTED** [review]
Overall looks good. Had a few suggestions for code style changes.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/BroadcastStateTransformation.java (hunk 4 above)
Comment (kl0u) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/BroadcastStateTransformation.java:54]: I like this comment. You are right!

**Comment by kl0u** [comment]
@dawidwys I integrated your comments, let me know what you think!

**Review by dawidwys — APPROVED** [review]
Thanks for the update! Looks good now.

+1 to merge, assuming azure is green ;)

**Comment by kl0u** [comment]
Thanks a lot! I will merge as soon as AZP gives green.

**Pull request closed without merging**